### ISSUE TITLE
support for "NONE" or "NOASSERTION" licenses

### ISSUE THREAD

**Issue by jandrieu** [npm/npm#8557]
SPDX provides for both "NONE" and "NOASSERTION" as valid license fields, but npm init accepts neither.

See http://spdx.org/sites/spdx/files/SPDX-2.0.pdf section Section 3.12-3.14

Is is related to https://github.com/npm/npm/issues/8542 and https://github.com/npm/npm/issues/8291

**Comment by kemitchell** [comment]
@jandrieu, thanks again for the issue, here and at kemitchell/spdx#11. I had second a look at the sections of the spec you referenced. Special thanks for calling them out by number.

I will let the SPDX team give the official word on the bug you opened at https://bugs.linuxfoundation.org/show_bug.cgi?id=1289, but will also go ahead and share my own reading, to keep things moving along.

Concluded license, info. from files, and declared license are part of the specification for "SPDX files", which are out of scope for [the spdx package](https://npmjs.com/packages/spdx). To be fair, that package is both narrower and broader in scope than the SPDX specification. It is narrower in that it implements only License Expressions (which in turn require the License Identifier and License Exception lists). It is broader in that it "overloads" the License Expression syntax as, first, a description of the license terms for packages, in line with the spec, and, second, an expression of license constraints. The latter semantics are my work, not set in stone, and very open to feedback. (See, e.g. kemitchell/spdx.js#8 from @phadej.) Any resemblance to [node-semver](https://npmjs.com/packages/semver) is very much intentional.

The SPDX file fields you reference also embody third-party assessments of the licensing terms for a given package. While there is nothing to stop a software project from including an SPDX file in source control or even publishing it, say, to npm, the primary use case is third-party assessments of "found object" software packages plucked from the world of open source.

This comes through in the semantics of `NONE` and `NOASSERTION`, which speak to a reviewer's failure to determine license terms due to lack of information, inconclusive information, or abstention. The same values mean much the same, with finer points appropriate in context, for many other SPDX file properties, such as 3.7 "Package Download Location" or 3.10 "Package Home Page".

It's entirely fair to point out that node 2.10.0 repurposes a relatively small part of the entire SPDX spec as a way for package authors to express license terms. (Blame RubyGems. I think they blazed that trail.) It's fair to point out that the legal effect of valid SPDX `license` property is dubious, even if it's a great sign practical risk is low. I think it's also fair to point out that @othiym23 and I moved very quickly, and that the one "magic value" we did approve, `LicenseRef-LICENSE`, utilizes a part of the License Expression syntax at the interface between License Expressions and the SPDX file format otherwise left behind.

I would not oppose a PR to add some other magic value, like `None`, to npm's validation routine, as a way to hush warnings. There is a valid use case in private modules. I will reserve the right, however, to heap Linus-level opprobrium on anyone dumb or insensitive enough to publish `None` in `package.json` to a public repository.

**Comment by kemitchell** [comment]
@jandrieu you may be interested to know that both SPDX legal and tech teams are working on revisions to `NONE` and `NOASSERTION`. I wasn't on the joint call this week, but I understand that the teams are discussing harmonizing the meanings of those values across SPDX fields:

http://lists.spdx.org/pipermail/spdx-tech/2015-June/002721.html

**Comment by jandrieu** [comment]
@kemitchell Very cool. I see the distinction you make and understand the unintended use of SPDX may mean some misalignment.

You did however, anticipate my initial thought, namely that while SPDX is about third party assessment, npm's package.json is about authors labeling their own work. 

As I just commented on the spdx.org issue, maybe the right answer is actually an explicit new valid license expression, e.g., "All rights reserved".

I'd be happy to make a PR if "NONE" is the right answer, but it seems to me the semantics are different in the npm use case and the SPDX standard.

It'd be great if the SPDX guys have (or had) some ideas on their call today. It would make more sense to follow their lead  than vary too much from that. However, if it makes sense to have a npm-specific value for a non-license, I think we should distinguish from NONE as currently defined in SPDX. Perhaps PRIVATE matches current package.json usage best.  

It would also make the heaped opprobrium all the more apt when someone publishes a private package to a public registry.

**Comment by kemitchell** [comment]
Great stuff!

Two quick points i would like to add:

First, I am on board with choosing a new magic value that does not evoke "NONE" or "NOASSERTION" from SPDX. I remain otherwise consciously indifferent to the choice of string.

But I'd like to be sensitive to npm's recent announcement of private modules. The existence of the private=true property is potential source of confusion enough. Perhaps license="Private Module" should hush the license warning, but warn on attempted public npm publish?

**Comment by jandrieu** [comment]
I suppose I share a general indifference to the specific term... even as I'm prone to elegant, meaningless semantic debates. Let'st do something simple. Maybe it would be good to avoid the term private at all if it brings in potential confusion with private modules--as you could have a private module (which is about public access to the source) with a specific license.

The notes after the SPDX call actually seemed to be less than clarifying. They proposed restricting NONE to just 3.14, but also "NONE should be used if there is no public (or internal) URL whatsoever."  And that confuses me more than it clarifies.

I did suggest a new license on the bugzilla issue, so we'll see if that gets picked up or not. If not, let's just pick something non-confrontational and use it.

**Comment by evangoer** [comment]
Thank you @jandrieu for creating this issue. If "NONE" or "NOASSERTION" turns out to not be a good option, then what about making the magic word "PROPRIETARY"? That might help avoid confusion with private modules? Apologies if I'm bikeshedding.

**Comment by othiym23** [comment]
`NONE` and `NOASSERTION` are very clear that they're not making any claims other than that there is, respectively, no license associated with a package, or that the package makes no assertion one way or another about its license. The problem with `PROPRIETARY` is that it implies that there exists a class of licenses with that name that share a set of characteristics, and I'm not sure how such a class could be meaningfully designated. It makes perfect sense from the packaging point of view, but remember the primary audience for SPDX assertions is the consumers, who need to be able to determine unambiguously whether a package or project is acceptable for use within their own applications.

That said, yeah, this is getting mildly bikesheddy, and not in the right venue – either the SPDX forums or kemitchell/spdx are the correct places to be talking about how this particular thing is spelled.

**Comment by kemitchell** [comment]
@othiym23, I think everyone involved has preemptively acknowledged that we stand at the the top of a steep slippery slope leading down to a bikeshed. We're stuck because nobody wants to be "the fool who spoke a color", not because strong opinions conflict. Nice problem to have.

Here's a proposal:

The bikeshed is picking magic values. We already have one, `LicenseRef-LICENSE`, an ugly hack born of expedience. (Mea culpa.) We need another that means more or less "not licensed". Forrest, brother, as "the npm guy" on the thread, I do humbly request that you pick pretty, non-hackish replacement values by fiat. For reasons above, my only humble request is that they _not_ be `NONE`, `NOASSERTION` or similar, to avoid confusion with the spec.

Since we're off the spec, these values shouldn't go in [spdx](https://npmjs.com/packages/spdx). The better candidate is probably [validate-npm-package-license](https://www.npmjs.com/package/validate-npm-package-license). The logic would end up looking psuedo-like:

```javascript
function validp(license) {
  if (spdx.valid(license)) {
    if (license.startsWith('LicenseRef')) {
      // Not using this for a magic value any more,
      // so this no longer makes sense outside the full SPDX context.
      return false;
    } else {
      return true;
    }
  } else if (license === 'MAGIC VALUE 1') {
    return true; // look at the license file
  } else if (license === 'MAGIC VALUE 2') {
    return true; // no license
  } else {
    return false;
  }
}
```

I'll be happy write the PRs when I can, since I already know my way through the weeds, but @jandrieu, please feel free to "call" some or any part of this that you want. I don't care about credit here---a certain someone already went way too far in release notes a while back---and you deserve some if you want it.

**Comment by kemitchell** [comment]
![9a0d7099f6825b02ef28ee090bae1e61](https://cloud.githubusercontent.com/assets/205760/8244983/f735d984-15dc-11e5-9400-4bfcc5881ae1.gif)

**Comment by othiym23** [comment]
* I don't like `"None"` because it's mostly the same as `"NONE"`. Overloading by case seems bad.
* A plain `false` (the value, not the string) is appealing, but that's just because it's notionally convenient to do that kind of type overloading in JSON, not because it's actually a good idea.
* Using a plain `null` would cause all kinds of havoc. Trust me on this.
* My objection to `"PRIVATE"` fits under the same heading as what I said to @evangoer about `"PROPRIETARY"`.

I think the plainest and most factual value for this would be simply `"UNLICENSED"` / `"UNLICENCED"` (gotta have both to preserve transAtlantic harmony). How does that grab everyone? It's sort of clunky, which is the point, but it's also a plain assertion on the part of the person publishing the package.

**Comment by kemitchell** [comment]
1. Case-sensitive?

2. Would you like to pick a value for "See the license file"? Since we're already defining another magic value, there's no reason to play games with the flexibility of `LicenseRef-*` any more. I will personally hunt down and PR any packages already using `LicenseRef-LICENSE`.

**Comment by othiym23** [comment]
1. I kinda like `"UNLICEN[SC]ED"` looking shouty, but I don't feel strongly about it. The rest of `spdx` appears to be case-sensitive, so it makes sense for this to follow suit.
2. I _like_ the `"LicenseRef-*"` mechanism, and would like to get it wired up in npm to actually verify the existence of the file thus specified and warn if it's missing (even if we don't do anything with its contents). If enough people used it, it would be useful to people writing automated license validators. Maybe those people would include npm! Who knows? The future is full of infinite possibility!

**Comment by evangoer** [comment]
"UNLICENSED" expresses the intent very clearly, as far as I'm concerned. Thank you, @othiym23!

**Comment by kemitchell** [comment]
1. 10-4. All caps.

2. I'm not suggesting that npm do without a value that means "refer to the license file", but that we use something like `SEE LICENSE FILE` instead of `LicenseRef-LICENSE`.

    The latter abuses `LicenseRef`by defining an npm-wide, rather than user-defined, special value. It also opens up a universe of valid but meaningless strings, since `LicenseRef-` followed by anything is a valid SPDX expression. `LicenseRef-LICENSE-BSD` and `LicenseRef-LICENCE`, for example.

    In other words, if we're going to address any nonstandard license scenarios with npm-specific magic, `LicenseRef` is a distant second best to magic, and can actually hurt consistency and automation. It will take a little code to make sure no `LicenseRef`s appear in any nested SPDX expressions, but those code paths will stay very, very cold.

**Comment by othiym23** [comment]
OK! I'm with you now, @kemitchell! How about `"SEE LICENSE IN <filename>"`?

**Comment by kemitchell** [comment]
```javascript
/^SEE LICENSE IN (.+)$/; // ?
```

**Comment by othiym23** [comment]
```javascript
/^SEE LICEN[SC]E IN (.+)$/ // ;)
```

**Comment by kemitchell** [comment]
Do not want to ask this:

```json
{
  "license":"MIT"
}
```

```json
{
  "licence":"MIT"
}
```

[normalize-package-data](https://www.npmjs.com/package/normalize-package-data), right?

**Comment by kemitchell** [comment]
@othiym23, are you planning to use the capture group to `stat` the file? Any concern re filesystem character support?

```javascript
/^SEE LICEN[SC]E IN ([A-Za-z0-9_]+)$/;
```

`(\w+)` would work, too, but the above is self-describing for those porting to different regular expression implementations.

**Comment by othiym23** [comment]
> Do not want to ask this:

`fstream-npm`, actually. npm only accepts the one name for the field in `package.json`, but npm is agnostic about the name of the license file, and that feels like a good practice (practise) to me. (Maybe we should accept both, and normalize to one, but that's out of scope for this chunk of the project).

> are you planning to use the capture group to stat the file?

Probably at some point that would be handy to have, although npm / `normalize-package-data` can parse it itself.

> Any concern re filesystem character support?

Postel's Law says no, we'll deal with that issue when and if we come to it.

**Comment by kemitchell** [comment]
re `fstream-npm`: Thanks for correcting. I will leave that to others.

re Postel: 10-4

**Comment by kemitchell** [comment]
A few dark thoughts.
```json
{"license":"UNLICENCED"}
```
:smirk:
```json
{"license":"SEE LICENCE FILE LICENSE.LICENCE"}
```
:see_no_evil:
```json
{"license":"SEE LICENSE FILE index.js"}
```
:fearful: 
```json
{"license":"SEE LICENSE FILE README.md"}
```
:smiling_imp:

I'll stop making thread-noise now.

**Comment by othiym23** [comment]
I know you're just funnin', but

> ```json
> {"license":"SEE LICENSE FILE README.md"}
> ```
> :smiling_imp:

seems perfectly legitimate to me -- before I spent all my days working on npm all the time, I commonly stuck my license at the bottom of README.md, and I think affording common use cases is a nice thing to do.

**Comment by kemitchell** [comment]
```bash
$ cat package.json | fgrep license
"license": "SEE LICENSE FILE README.md",

$ tail -n 2 README.md
## License
MIT License

$ echo -n "SEE LICENSE IN README.md" >> ~/.npmrc
```

**Comment by othiym23** [comment]
Closed by 16a3dd545b10f8a2464e2037506ce39124739b41, which is the keystone to supporting `UNLICENSED` and `SEE LICENSE IN <file>`. Thanks to everyone for the thoughtful discussion!

**Comment by jandrieu** [comment]
Thanks for the prompt fix. That was impressive.

**Comment by alexindigo** [comment]
@othlym23 Is it possible to just turn off warning with one of the npm options? Thank you.

**Comment by jandrieu** [comment]
This should no longer be closed.  The "prompt" fix has been reverted for some reason.

**Comment by e-oz** [comment]
All issues are closed but it's still not clear what to write in case of private code to suppress warning.

**Comment by evangoer** [comment]
Hi @e-oz -- For private code, the terse way to do it is to say "UNLICENSED". See documentation at https://docs.npmjs.com/files/package.json.

**Comment by kirkstrobeck** [comment]
Just add `"private": true,`, then you need no license entry

**Comment by jandrieu** [comment]
@kirkstrobeck You can't do that during the interactive ```npm init```.

**Comment by rsp** [comment]
@jandrieu @e-oz @kirkstrobeck is there any consensus on how proprietary code should be marked as such?

Issues #8542 got closed as a duplicate of this one and #8291, this one is also closed by 16a3dd5 which added "UNLICENSED" and "SEE LICENSE IN <filename>" but people are still recommending to use `"private": true` instead of that?

As for me, I'd like to see "PROPRIETARY" added to the list for proprietary licenses, since UNLICENSED seems inaprorpiate for cases when there is a license but not an open-sorce one but since this is not an option then my question is which of the 4 options is most common and/or preferred:

1. "license": "UNLICENSED"
2. "license": "SEE LICENSE IN LICENSE.md"
3. "license": "UNLICENSED", "private": true
4. "license": "SEE LICENSE IN LICENSE.md", "private": true
5. "private": true (with no "license")

I'm asking because reading discussions on all of the relevant issues it doesn't seam clear what the consensus is, if there is any.

**Comment by kirkstrobeck** [comment]
I use number 5

> "private": true (with no "license")

**Comment by alexindigo** [comment]
In our company we're using `1.` and it seems to be clear for developers and non-developers.
Also works well with license scanners.

**Comment by evangoer** [comment]
As @othiym23 pointed out to me near the beginning of this thread, "UNLICENSED" is not the same thing as proprietary. "UNLICENSED" is a machine-readable value that says, you haven't given your software a license. Proprietary means, you're using a non-standard license -- for example, "You may freely use my software if you post a video of yourself playing 'Twinkle Twinkle Little Star' on the violin."

For proprietary licenses, I would use option 2.

**Comment by alexindigo** [comment]
And in most internal-only packages, you’re not licensing (not releasing) your code, you’re keeping it “in house”. So it’s unlicensed.
